darwin/importer/importer.py

```diff
@@ -1472,3 +1500,168 @@ def _overwrite_warning(
+
+
+def _get_annotation_format(
+    importer: Callable[[Path], Union[List[dt.AnnotationFile], dt.AnnotationFile, None]]
+) -> str:
+    """
+    Returns the annotation format of the importer used to parse local annotation files
+
+    Parameters
+    ----------
+    importer : Callable[[Path], Union[List[dt.AnnotationFile], dt.AnnotationFile, None]]
+        The importer used to parse local annotation files
+
+    Returns
+    -------
+    annotation_format : str
+        The annotation format of the importer used to parse local files
+    """
+    return importer.__module__.split(".")[3]
+
+
+def _verify_slot_annotation_alignment(
+    local_files: List[dt.AnnotationFile],
+    remote_files: Dict[str, Dict[str, Any]],
+) -> Tuple[List[dt.AnnotationFile], Dict[str, List[str]], Dict[str, List[str]]]:
+    """
+    Runs slot alignment validation against annotations being imported. The following checks are run:
+    - For multi-slotted items:
+        - For every annotation not uploaded to a specific slot:
+          A non-blocking warning is generated explaining that it will be uploaded to the default slot
+
+    - For multi-channeled items:
+        - For every annotation not uploaded to a specific slot:
+          A non-blocking warning is generated explaining that it will be uploaded to the base slot
+        - For every annotation uploaded to a slot other than the base slot:
+          A blocking error is generated explaining that annotations can only be uploaded to the base slot of multi-channeled items
+
+    Files that generate exclusively non-blocking warnings will have those warnings displayed, but their import will continue.
+    Files that generate any blocking error will only have their blocking errors displayed, and they are removed from the list of files to be imported.
+
+    Errors can only be generated by referring to multi-slotted or multi-channeled items.
+    These concepts are only supported by Darwin JSON 2.0, so stop imports of other formats if any warnings occur.
+
+    Parameters
+    ----------
+    local_files : List[dt.AnnotationFile]
+        A list of local annotation files to be uploaded
+    remote_files : Dict[str, Dict[str, Any]]
+        Information about each remote dataset item that corresponds to the local annotation file being uploaded
+
+    Returns
+    -------
+    local_files : List[dt.AnnotationFile]
+        A pruned list of the input annotation flies. It excludes any input files that generated a blocking warning
+    slot_errors : Dict[str, List[str]]
+        A dictionary of blocking errors for each file
+    slot_warnings : Dict[str, List[str]]
+        A dictionary of non-blocking warnings for each file
+    """
+
+    slot_errors, slot_warnings = {}, {}
+    for local_file in local_files:
+        remote_file = remote_files[local_file.full_path]
+        local_file_path = str(local_file.path)
+        if len(remote_file["slot_names"]) == 1:
+            continue  # Skip single-slotted items
```

Comment:
While working on this, I noticed another case we could check against. When uploading annotations to any item (even single-slotted ones), it's possible to upload them to a slot that doesn't exist. In this case they simply don't appear in the workview, but these "black holed" annotations are in exports of the item. Should we block this?

If so, I can change this section and add that validation here